### PR TITLE
CRM457-2122: Improve autocomplete effectiveness

### PR DESCRIPTION
**Pull request by elken**
## Description of change

[Link to relevant ticket](https://dsdmoj.atlassian.net/browse/CRM457-2122)

This PR aims to improve the autocomplete in a number of ways:

### Ignore spaces
Sometimes users put a space at the start or the end by mistake, as such leading and trailing spaces should be trimmed.

### Ignore case
Case here is only useful to display grammatically correct strings to the user. A&E and a&e are identical so should be treated as such

### Ignoring stop words
The way this is handled should probably change, the hardcoded list is somewhat inflexible _but_ I also don't imagine the list will be changed much once it's committed.

### Fuzzy matching
Candidates should also be matched regardless of word order and regardless of if all the words are present. The following cases are now considered

- User types "Child Psychologist" and both that and "Psychiatrist (including child)" are returned
- User types "Forensic Firearm" and both "Forensic scientist" and "firearm expert" are returned
- User types "Consultant Forensic psychiatrist" and "Forensic psychiatrist" is returned

### `autoselect: true`
This will prevent users from saving custom entries when they instead mean to just select the first entry when pressing `Return`.

### Sorting by frequency
Matches will first attempt to sort by how often they occur if the autocomplete component has the data passed as `data-values`.

### String distance
Matches will be sorted by how "distant" they are from the query using [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) after attempting to sort by frequency. This won't always be correct, but close matches should be right near the top.

### Ignore punctuation
Candidates that have ampersands and quotes should match even without the user putting them in

### CODE REVIEW

**Review by patrick-laa — COMMENTED** [review]
So first of all this is awesome. I had a quick play on the dev branch and it behaves really nicely.

BUT I think quite a lot of the problems we need to solve are server-side. E.g. the problem described in improvement 2 is about case-sensitivity, where someone has written 'accountant' and we've not associated that with the 'Accountant' service type. Now, the auto-complete is already case sensitive - if you type 'accountant' you'll see 'Accountant' in the dropdown. So the problem appears to be that providers aren't clicking on Accountant even though we know they see it. The solution therefore has to be some server-side intelligence so that if the `service_type_autocomplete_suggestion` matches a known value not-quite-exactly we still associate the quote with that known value. I think this means, in `PriorAuthority::Steps::PrimaryQuoteForm#service_type_autocomplete_suggestion=(value)`, trimming whitespace, adding case-insensitivity, and ignoring stop words. We don't want to match _too_ strongly in case a provider has deliberately put in something custom which we don't want to throw away, but the "this is clearly what we meant" cases are ones we can definitely fix.

**Comment by elken** [comment]
> So first of all this is awesome. I had a quick play on the dev branch and it behaves really nicely.
> 
> BUT I think quite a lot of the problems we need to solve are server-side. E.g. the problem described in improvement 2 is about case-sensitivity, where someone has written 'accountant' and we've not associated that with the 'Accountant' service type. 

The problem is the searches _are_ case sensitive, the solution is they should _not_ be case-sensitive. Coupled with `autoselect: true` it's now really hard to mistype a candidate this way. Will need a BA to clarify these requirements further though if there's uncertainty here.

> Now, the auto-complete is already case sensitive - if you type 'accountant' you'll see 'Accountant' in the dropdown. 
So the problem appears to be that providers aren't clicking on Accountant even though we know they see it. The solution therefore has to be some server-side intelligence so that if the `service_type_autocomplete_suggestion` matches a known value not-quite-exactly we still associate the quote with that known value. I think this means, in `PriorAuthority::Steps::PrimaryQuoteForm#service_type_autocomplete_suggestion=(value)`, trimming whitespace, adding case-insensitivity, and ignoring stop words. We don't want to match _too_ strongly in case a provider has deliberately put in something custom which we don't want to throw away, but the "this is clearly what we meant" cases are ones we can definitely fix.

I'm not convinced here, for all my testing these client-side changes are enough to mitigate against a lot of the list Mary shared with us (I didn't do all of them) barring the synonyms that I am reminded to ticket up still.

It's a lot harder to do the wrong thing, and for the average user that's probably enough.

File: app/javascript/autocomplete.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { distance } from "fastest-levenshtein";
+import accessibleAutocomplete from "accessible-autocomplete";
+
+// Convert select fields to use accessible-autocomplete instead.
+//
+// Also includes functions to support requirements to make the autocompletion more fuzzy
+const stopWords = ["a", "the", "of", "in", "and", "at", "&"];
```
[the remark this answers is not in the document]
Comment (elken) [reply app/javascript/autocomplete.js:7]: Also noticed I need to handle only ignoring stop words if there's other words in the query, "a" returns nothing

**Comment by patrick-laa** [comment]
Ah, apologies! In my playing around I hadn't noticed that, as you say:
> Coupled with `autoselect: true` it's now really hard to mistype a candidate this way.

This completely solves the thing I was worried about. Marvellous.

**Comment by patrick-laa** [comment]
I do note that the autoselect is quite strong. If any of the words in what you type matches any of the words in any of the autocomplete entries, you cannot add a custom entry. E.g. I cannot enter "Arborist's report" because it will _always_ adjust it to the closest match of all the other "X report" entries, even though none of them are anything like Arborist. I can imagine someone who really does want to write something custom being pretty frustrated with this, especially if it silently selects a different service type without them noticing. But this may be more a UX decision.

**Comment by elken** [comment]
> I do note that the autoselect is quite strong. If any of the words in what you type matches any of the words in any of the autocomplete entries, you cannot add a custom entry. E.g. I cannot enter "Arborist's report" because it will _always_ adjust it to the closest match of all the other "X report" entries, even though none of them are anything like Arborist. I can imagine someone who really does want to write something custom being pretty frustrated with this, especially if it silently selects a different service type without them noticing. But this may be more a UX decision.

It's a valid point, but unfortunately the requirements of the ticket do remove the ability for custom things like that to match.

The alternative is we remove `autoselect: true` and maybe add some hint text, but as you say this is a UX point. I'll raise it with Mary on the autoselect spike ticket.

**Comment by elken** [comment]
Ah Mary has already raised this on said ticket and we're getting rid of the `autoselect: true`

**Comment by patrick-laa** [comment]
> Ah Mary has already raised this on said ticket and we're getting rid of the autoselect: true

Ah ok. Annoyingly that kinda revives my "let's do some stuff server-side to cover when a user types something that, case-insentively and whitespaced-removed, matches something in the dropdown, but they don't click on the dropdown" thing.

**Comment by elken** [comment]
Uh apparently I edited your comment instead of adding one? THanks Github??

The client takes care of all of that still, the only thing not considered is the user not selecting anything but there's a few other options I can look at that still allow custom ones not in the list.

I have an idea that might allow autoselect: true and custom ones, but I'll need to look at it.

File: app/views/prior_authority/steps/primary_quote/edit.html.erb [hunk 2]
```diff
@@ -14,7 +14,7 @@
       <% if @form_object.draft? %>
         <%= suggestion_select form, :service_type_autocomplete, PriorAuthority::QuoteServices.values, :to_sym, :translated, width: 20,
```
Comment (robinsk1) [inline app/views/prior_authority/steps/primary_quote/edit.html.erb:15]: We already call PriorAuthority::QuoteServices.values in the controller - do we need to call that twice?

**Review by robinsk1 — COMMENTED** [review]
💯

File: app/controllers/prior_authority/steps/primary_quote_controller.rb [hunk 3]
```diff
@@ -2,6 +2,15 @@ module PriorAuthority
...
       def edit
+        counts = PriorAuthorityApplication
+                 .where.not(service_type: [nil, ''])
+                 .group(:service_type)
+                 .count
+        values = PriorAuthority::QuoteServices.values.map do |service|
+          [service.translated, counts.fetch(service.value.to_s, 0)]
+        end
+
+        @values = values.sort_by { |_, count| -count }.to_h
```
[the remark this answers is not in the document]
Comment (elken) [reply app/controllers/prior_authority/steps/primary_quote_controller.rb:13]: Sure, slightly clearer

File: app/controllers/prior_authority/steps/primary_quote_controller.rb [hunk 4]
```diff
@@ -2,6 +2,15 @@ module PriorAuthority
   module Steps
     class PrimaryQuoteController < BaseController
       def edit
+        counts = PriorAuthorityApplication
```
[the remark this answers is not in the document]
Comment (elken) [reply app/controllers/prior_authority/steps/primary_quote_controller.rb:5]: Could do, it's not going to be used anywhere else but easy enough to refactor

File: app/views/prior_authority/steps/primary_quote/edit.html.erb (hunk 2 above)
Comment (elken) [reply app/views/prior_authority/steps/primary_quote/edit.html.erb:15]: Probably not, but it's a static call so I didn't think it was worth changing

**Review by patrick-laa — APPROVED** [review]

**Pull request merged**